beacon-chain/blockchain/chain_info.go

```diff
@@ -141,46 +139,39 @@ func (s *Service) HeadRoot(ctx context.Context) ([]byte, error) {
 
 // HeadBlock returns the head block of the chain.
 func (s *Service) HeadBlock() *ethpb.SignedBeaconBlock {
-	s.headLock.RLock()
-	defer s.headLock.RUnlock()
-
-	return proto.Clone(s.headBlock).(*ethpb.SignedBeaconBlock)
+	return s.headBlock()
 }
 
 // HeadState returns the head state of the chain.
 // If the head state is nil from service struct,
 // it will attempt to get from DB and error if nil again.
 func (s *Service) HeadState(ctx context.Context) (*state.BeaconState, error) {
-	s.headLock.RLock()
-	defer s.headLock.RUnlock()
-
-	if s.headState == nil {
-		headState, err := s.beaconDB.HeadState(ctx)
-		if err != nil {
-			return nil, err
-		}
-		s.headState = headState
-		return headState, nil
+	if s.hasHeadState() {
+		return s.headState(), nil
 	}
 
-	return s.headState.Copy(), nil
+	headState, err := s.beaconDB.HeadState(ctx)
```

Comment:
Update cache if we have a cache miss?

Reply:
Not here. That's what the PR aim to avoid. Update individual head field. If there's a miss `setHead` will be used during init